nav2_costmap_2d/plugins/obstacle_layer.cpp

```diff
@@ -313,7 +321,77 @@ ObstacleLayer::dynamicParametersCallback(
       if (param_name == name_ + "." + "combination_method") {
         combination_method_ = parameter.as_int();
       }
+    } else if (param_type == ParameterType::PARAMETER_STRING) {
+      if (param_name == name_ + "." + "observation_sources") {
+        std::stringstream ss(parameter.as_string());
+        std::string source;
+
+        std::stringstream ss_old(observation_sources_);
+        std::string source_old;
+
+        bool source_exists = false;
+        std::vector<std::string> matched_sources;
+        std::ostringstream new_sources_stream;
+
+        //find all the sources that have already been set.
+        while (ss >> source) {
+          while (ss_old >> source_old) {
+            if (source == source_old) {
+              source_exists = true;
+              matched_sources.push_back(source);
+              break;
+            }
+          }
+
+          if (!source_exists) {
+            if (!new_sources_stream.str().empty()) {
+              new_sources_stream << " ";
+            }
+            new_sources_stream << source;
+            continue;
+          }
+
+          source_exists = false;
+        }
+
+        std::string new_sources = new_sources_stream.str();
+        std::string topic;
+        std::int16_t index = 0;
+        std::stringstream ss_old_final(observation_sources_);
+
+        // find and remove the sources that were not passed into the parameter update
+        while (ss_old_final >> source_old) {
+          if (std::find(matched_sources.begin(), matched_sources.end(), source_old) == matched_sources.end()) {
+            // The observation source needs to be removed
+            RCLCPP_INFO(
+              logger_,
+              "Removing observation source: %s", source_old.c_str());
+
+            // assume that the order of the sources in observation_sources_ is the same as in the vectors
```

Comment:
Make sure these are in the right order to not cause segfaults or accessing incorrect memory. Also make sure to unsubscriber or otherwise reset the state before destroying them to also prevent other potential race conditions